pykafka/pykafka/producer.py

```diff
@@ -69,43 +71,70 @@ def __init__(self,
 
-    def _send_request(self, broker, req):
-        tries = 0
-        while tries < self._max_retries:
-            try:
-                broker.produce_messages(req)
-                break
-            except (UnknownTopicOrPartition, LeaderNotAvailable,
-                    NotLeaderForPartition, RequestTimedOut) as ex:
-                # FIXME: Not all messages will have failed. Only some on a
-                #        bad partition. Retrying should reflect that. The
-                #        fix needs to be in protocol.py as well since it
-                #        only raises one exception no matter how many errors.
-                tries += 1
-                if tries >= self._max_retries:
-                    raise
-                elif isinstance(ex, UnknownTopicOrPartition):
-                    logger.warning('Unknown topic: %s. Retrying.', self._topic)
-                elif isinstance(ex, LeaderNotAvailable):
-                    logger.warning('Partition leader unavailable. Retrying.')
-                elif isinstance(ex, NotLeaderForPartition):
-                    # Update cluster metadata and retry the produce request
-                    # FIXME: Can this recurse infinitely?
-                    # FIXME: This will cause a re-partitioning of messages
-                    self._cluster.update()
-                    self._produce(req.messages)
-                elif isinstance(ex, RequestTimedOut):
-                    logger.warning('Produce request timed out. Retrying.')
-
-    def _produce(self, messages):
-        """Publish a set of messages to relevant brokers."""
-        # Requests grouped by broker
-        requests = defaultdict(lambda: ProduceRequest(
-            compression_type=self._compression,
-            required_acks=self._required_acks,
-            timeout=self._ack_timeout_ms
-        ))
+    def _send_request(self, broker, req, attempt):
+        """Send the request to the broker and handle the response."""
+
+        def _get_partition_msgs(partition_id, req):
+            """Get all the messages for the partitions from the request."""
+            messages = itertools.chain.from_iterable(
+                mset.messages
+                for topic, partitions in req.msets.iteritems()
+                for p_id, mset in partitions.iteritems()
+                if p_id == partition_id
+            )
+            for message in messages:
+                yield (message.partition_key, message.value), partition_id
+
+        # Do the request
+        to_retry = []
+        try:
+            response = broker.produce_messages(req)
+
+            # Figure out if we need to retry any messages
+            to_retry = []
+            for topic, partitions in response.topics.iteritems():
+                for partition, (error, offset) in partitions.iteritems():
+                    if error == 0:
+                        continue  # All's well
+                    if error == UnknownTopicOrPartition.ERROR_CODE:
```

Comment:
I think it's fine for this pull request - we may end up realizing during testing that this is wrong.